### PR TITLE
Deleting characters

### PR DESCRIPTION
**Pull request by AnywayFarus**
## About The Pull Request

Adds a button to delete characters, and when deleted it flips to the nearest character, wow(the last character can't be deleted(I guess)).

### Video: 

https://github.com/tgstation/tgstation/assets/78199449/3ad1157a-6c1c-4fb2-9b0d-1da791cad989

## Why It's Good For The Game

Once you know how to create characters, you sometimes want to be able to delete them.
The reasons for this can be... as simple as freeing up a slot, or deleting them in order to remake them.
I think this is quite a useful feature that should have been there before.

## Changelog

:cl: Vishenka0704
qol: Ability to delete characters(yourself)
/:cl:

### CODE REVIEW

File: code/modules/client/preferences.dm [hunk 1]
```diff
@@ -224,6 +224,22 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 			character_preview_view.update_body()
 
 			return TRUE
+		if ("remove_slot")
+			var/pickedSlot = params["slot"]
+			var/confidence_check = tgui_input_text(usr, "To confirm the deletion of a character, type the name of the character.", "Character Deletion")
+			if(confidence_check == params["name"])
```
Comment (san7890) [inline code/modules/client/preferences.dm:230]: turn this into an early return so we don't have to worrk about indentation here

```dm
if(confidence_check != params["name"])
 return FALSE
```

File: code/modules/client/preferences.dm [hunk 2]
```diff
@@ -224,6 +224,22 @@ GLOBAL_LIST_EMPTY(preferences_datums)
...
+				if(remove_character(pickedSlot))
```
Comment (san7890) [inline code/modules/client/preferences.dm:231]: this can also be an early `return FALSE`.

File: code/modules/client/preferences.dm [hunk 3]
```diff
@@ -224,6 +224,22 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 			character_preview_view.update_body()
 
 			return TRUE
+		if ("remove_slot")
+			var/pickedSlot = params["slot"]
```
Comment (san7890) [inline code/modules/client/preferences.dm:228]:
```suggestion
			var/picked_slot = params["slot"]
```

prefer `snake_case`

**Review by san7890 — CHANGES_REQUESTED** [review]

**Review by MrMelbert — CHANGES_REQUESTED** [review]
It's kind of a weird place to put a delete character button, leaves a bunch of empty space

The desert
image

At the very least I would keep the button always-shown and just disable it when it's unusable.

**Comment by MrMelbert** [comment]
@Mothblocks Got an opinion on where it should go?

**Comment by optimumtact** [comment]
Does this button have a safety feature?

it should make you type the character name.

**Review by ZephyrTFA — APPROVED** [review]

**Comment by AnywayFarus** [comment]
> It's kind of a weird place to put a delete character button, leaves a bunch of empty space
> 
> The desert image
> 
> At the very least I would keep the button always-shown and just disable it when it's unusable.

image

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
see above

**Comment by ZephyrTFA** [comment]
if your justification is the downstream wants it.
publish it downstream

**Comment by AnywayFarus** [comment]
> if your justification is the downstream wants it. publish it downstream

But no one above, except you two, has said you don't need this feature.
I came to you, not because you have to keep maintaining it, but in case you need it, that's how opensource works, no?

**Comment by AnywayFarus** [comment]
> Your justification is bad. Why should we maintain this if we don't want it. We are not a library. Also why does it not update the ui on delete in your video.

Well, as you coded, that's how it works :).
Manual update used to break styles (and rarely even the whole pref-menu).

**Comment by mc-oofert** [comment]
I use 1 slot

**Comment by ZephyrTFA** [comment]
> > if your justification is the downstream wants it. publish it downstream
> 
> But no one above, except you two, has said you don't need this feature.
> I came to you, not because you have to keep maintaining it, but in case you need it, that's how opensource works, no?

Your WIGFTG clearly says you are adding this for downstreams.

**Comment by AnywayFarus** [comment]
> > > if your justification is the downstream wants it. publish it downstream
> > 
> > 
> > But no one above, except you two, has said you don't need this feature.
> > I came to you, not because you have to keep maintaining it, but in case you need it, that's how opensource works, no?
> 
> Your WIGFTG clearly says you are adding this for downstreams.
All the features that I've added here are necessary for me specifically, only on downstreams.
But I'm sharing them on the premise that you might find it useful.

**Comment by LemonInTheDark** [comment]
> But no one above, except you two, has said you don't need this feature. I came to you, not because you have to keep maintaining it, but in case you need it, that's how opensource works, no?

I'm not looking to close this pr, I just want you to justify it in a way that is relevant to this codebase. I do not think the idea is inherently bad but your justification read as "This likely isn't useful to you but it is for me". I'm not sure if that was intended though.

You need some reason for us to potentially want this. Since you don't play here the reason you want it for downstream would work, it's possible that would fit some aspect of our project.

**Comment by LemonInTheDark** [comment]
> if your justification is the downstream wants it. publish it downstream

I do not think this was justification to close, just justification to block since it's something we can discuss and resolve.

> I don't like that we are setting precedents here where I feel the need to warn folks against using the magic word 'downstream' whenever they try to make a PR here.

If someone makes a pr here (or an issue report) it needs to be in the context of this project. I get many people come from downstream but if they're justifying something they need to justify it for THIS codebase. That's it. Can mention a downsteam while also doing that.

**Comment by LemonInTheDark** [comment]
Spoke to zypher, reopening conditional on some new justification

**Comment by LemonInTheDark** [comment]
> Cool. I don't mean to be a pain about this, but it's just something I noticed and felt the need to speak up about because it seemed a bit over the top. And hate to see another potential regular contrib be scared off over something silly

If it makes you feel better it was not because of what you said

**Comment by san7890** [comment]
the value is apparent for our codebase anyways since it's easier to start from scratch on rebalancing quirks or inventory loadouts. i understand the pr author needs to make those changes but i still have interest in this PR - would have merged it already if it weren't for some concerns on the UI layout

**Comment by AnywayFarus** [comment]
Uh-oh-oh.
I've gotten better results already on the downstream itself, I'll try to transfer a similar thing here as well.
It's just a different format. of the character list, you have buttons and they have a drop-down list.
I found a solution for auto-updating... which kind of improves the interaction itself a hundredfold, and I think it would fit in here too.

However, I haven't figured out a better place for the TG than the button at the bottom.

**Comment by AnywayFarus** [comment]
And in general, after translation (or maybe in your language originally) it looked quite rude, like, I offered you qol, instead of saying that "thank you, but we need", they say "why the hell are you trying to put it in our hands".

**Comment by AnywayFarus** [comment]
Updated the functionality itself, updated the video with the current functionality, updated WIGFTG

**Pull request merged**